src/emqx_access_control.erl

```diff
@@ -91,10 +89,29 @@ inc_acl_metrics(cache_hit) ->
     emqx_metrics:inc('client.acl.cache_hit').
 
 %% Auth
-default_auth_result(Zone) ->
-    case emqx_zone:get_env(Zone, allow_anonymous, false) of
-        true  -> #{auth_result => success, anonymous => true};
-        false -> #{auth_result => not_authorized, anonymous => false}
+default_auth_result(Username, Zone) ->
+    IsAnonymous = (Username =:= undefined orelse Username =:= <<>>),
+    AllowAnonymous = emqx_zone:get_env(Zone, allow_anonymous, false),
+    Bypass = emqx_zone:get_env(Zone, bypass_auth_plugins, false),
+    %% the `anonymous` filed in auth result does not mean the client is
+    %% connected without username, but if the auth result is based on
+    %% allowing anonymous access.
+    IsResultBasedOnAllowAnonymous =
+        case AllowAnonymous of
+            true -> true;
+            _ -> false
+        end,
+    Result = case AllowAnonymous of
+                 true -> #{auth_result => success, anonymous => IsResultBasedOnAllowAnonymous};
+                 _ -> #{auth_result => not_authorized, anonymous => IsResultBasedOnAllowAnonymous}
+             end,
```

Comment:
M.b. just 🤔 
```erlang
    Result = case AllowAnonymous of
                 true -> #{auth_result => success, anonymous => true};
                 _ -> #{auth_result => not_authorized, anonymous => false}
             end,
```

Reply:
use a more readable var name for document purpose.